shared/src/types/ethereum_events/vote_extensions.rs

```diff
@@ -0,0 +1,299 @@
+            ));
+        }
+        Ok(Self(ratio))
+    }
+}
+
+impl From<&FractionalVotingPower> for (u64, u64) {
+    fn from(ratio: &FractionalVotingPower) -> Self {
+        (ratio.0.numer().to_owned(), ratio.0.denom().to_owned())
+    }
+}
+
+impl BorshSerialize for FractionalVotingPower {
+    fn serialize<W: ark_serialize::Write>(
+        &self,
+        writer: &mut W,
+    ) -> std::io::Result<()> {
+        let (numer, denom): (u64, u64) = self.into();
+        (numer, denom).serialize(writer)
+    }
+}
+
+impl BorshDeserialize for FractionalVotingPower {
+    fn deserialize(buf: &mut &[u8]) -> std::io::Result<Self> {
+        let (numer, denom): (u64, u64) = BorshDeserialize::deserialize(buf)?;
+        Ok(FractionalVotingPower(Ratio::<u64>::new(numer, denom)))
+    }
+}
+
+impl BorshSchema for FractionalVotingPower {
+    fn add_definitions_recursively(
+        definitions: &mut std::collections::HashMap<
+            borsh::schema::Declaration,
+            borsh::schema::Definition,
+        >,
+    ) {
+        let fields =
+            borsh::schema::Fields::UnnamedFields(borsh::maybestd::vec![
+                u64::declaration(),
+                u64::declaration()
+            ]);
+        let definition = borsh::schema::Definition::Struct { fields };
+        Self::add_definition(Self::declaration(), definition, definitions);
+    }
+
+    fn declaration() -> borsh::schema::Declaration {
+        "FractionalVotingPower".into()
+    }
+}
+
+/// Aggregates an Ethereum event with the corresponding
+// validators who saw this event.
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize, BorshSchema)]
+pub struct MultiSignedEthEvent {
+    /// The Ethereum event that was signed.
+    pub event: EthereumEvent,
+    /// List of addresses of validators who signed this event
+    pub signers: HashSet<Address>,
+}
+
+/// Compresses a set of signed `VoteExtension` instances, to save
+/// space on a block.
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize, BorshSchema)]
+pub struct VoteExtensionDigest {
+    /// The signatures and signing address of each VoteExtension
+    pub signatures: Vec<(Signature, Address)>,
+    /// The events that were reported
+    pub events: Vec<MultiSignedEthEvent>,
+    /// The validators who saw no events
+    pub nulls: HashSet<Address>,
+}
+
+impl VoteExtensionDigest {
+    /// Decompresses a set of signed `VoteExtension` instances.
+    pub fn decompress(
+        self,
+        last_height: BlockHeight,
+    ) -> Vec<Signed<VoteExtension>> {
+        let VoteExtensionDigest {
+            signatures,
+            events,
+            nulls,
+        } = self;
+
+        let mut extensions = vec![];
+
+        for (sig, addr) in signatures.into_iter() {
+            let mut ext = VoteExtension::empty(last_height);
+
+            // TODO: perhaps remove the `nulls` field,
+            // as this code will behave much the same without it
+            if !nulls.contains(&addr) {
```

Comment:
You're completely right. We should remove the nulls field. I would rather have slightly longer execution times if it means a smaller digest.